codersdk/agentsdk/logs.go

```diff
@@ -483,6 +483,34 @@ func (l *LogSender) GetScriptLogger(logSourceID uuid.UUID) ScriptLogger {
 	return ScriptLogger{srcID: logSourceID, sender: l}
 }
 
+// WaitUntilEmpty waits until the LogSender's queues are empty or the given context expires.
+func (l *LogSender) WaitUntilEmpty(ctx context.Context) error {
+	ctxDone := false
+	nevermind := make(chan struct{})
+	defer close(nevermind)
+	go func() {
+		select {
+		case <-ctx.Done():
+			l.L.Lock()
+			defer l.L.Unlock()
+			ctxDone = true
+			l.Broadcast()
+			return
+		case <-nevermind:
+			return
+		}
+	}()
```

Comment:
The problem with an `allSent` channel is how to arrange when it should read.  Closing the channel won't work, because you can't "unclose" it if more data gets queued.

Writing to the channel won't work if there are more than one caller to WaitUntilEmpty.

Reply:
True, it would work better/simpler if the send loop was channel-based as well. In that case, one approach could be this:

```go
func (l *LogSender) WaitUntilEmpty(ctx context.Context) error {
	wait := make(chan struct{})
	l.waitUntilEmpty <- wait

	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-wait:
		return nil
	}
}

// SendLoop
	var waiters []chan struct{}
	for {
		select {
		case <-tick:
		case <-l.waitUntilEmpty:
			waiters = append(waiters, wait)
		}

		// ...

		if len(l.queues) == 0 {
			for _, wait := range waiters {
				close(wait)
			}
			waiters = nil
		}
```

But it's not quite as nice when retrofitted into the mutex style loop.

Reply:
The problem there is that it requires `SendLoop` to actually be running in order for `WaitUntilEmpty()` to return, which it might not be but we are still empty.

Channels are great for communicating between goroutines.  Here what we really, actually want is to know when a _condition_ is satisfied, regardless of other running goroutines, and for that `sync.Cond` is your friend.